staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting/validation.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package defaulting
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/validate"
+
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	schemaobjectmeta "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta"
+	"k8s.io/apiextensions-apiserver/pkg/apiserver/schema/pruning"
+	apiservervalidation "k8s.io/apiextensions-apiserver/pkg/apiserver/validation"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+// ValidateDefaults checks that default values validate and are properly pruned.
+func ValidateDefaults(pth *field.Path, s *structuralschema.Structural) (field.ErrorList, error) {
+	f := NewRootObjectFunc().WithTypeMeta(metav1.TypeMeta{APIVersion: "validation/v1", Kind: "Validation"})
+	v := validator{s}
+	return v.validate(pth, s, f, false)
+}
+
+type validator struct {
+	rootSchema *structuralschema.Structural
+}
+
+// validate is the recursive step func for the validation. insideMeta is true if s specifies TypeMeta or ObjectMeta.
+func (v *validator) validate(pth *field.Path, s *structuralschema.Structural, f SurroundingObjectFunc, insideMeta bool) (field.ErrorList, error) {
+	if s == nil {
+		return nil, nil
+	}
+
+	if s.XEmbeddedResource {
+		insideMeta = false
+		f = NewRootObjectFunc().WithTypeMeta(metav1.TypeMeta{APIVersion: "validation/v1", Kind: "Validation"})
+		v2 := *v
+		v = &v2
+		v.rootSchema = s
+	}
+
+	allErrs := field.ErrorList{}
+
+	if s.Default.Object != nil {
+		validator := validate.NewSchemaValidator(s.ToGoOpenAPI(), nil, "", strfmt.Default)
+
+		if insideMeta {
+			obj, _, err := f(runtime.DeepCopyJSONValue(s.Default.Object))
+			if err != nil {
+				return nil, fmt.Errorf("failed to prune default value: %v", err)
+			}
+
+			// check ObjectMeta/TypeMeta and everything else
+			if err := schemaobjectmeta.Coerce(nil, obj, v.rootSchema, true, false); err != nil {
+				allErrs = append(allErrs, field.Invalid(pth.Child("default"), s.Default.Object, fmt.Sprintf("must result in valid metadata: %v", err)))
+			} else if errs := schemaobjectmeta.Validate(nil, obj, v.rootSchema, true); len(errs) > 0 {
```

Comment:
what would this line do with this schema:
```yaml
type: object
properties:
  foo:
    type: object
    x-kubernetes-embedded-resource: true
    x-kubernetes-preserve-unknown-fields: true
    required: ["spec"]
    properties:
      metadata:
        type: object
        default:
          labels:
            foo: bar
      apiVersion:
        type: string
          enum: ["apps/v1"]
      kind:
        type: string
          enum: ["Deployment"]
      spec:
        type: object
        x-kubernetes-preserve-unknown-fields: true
```

if I'm reading this correctly, it *seems* like it would try to validate the following against the `foo` property's schema:
```yaml
apiVersion: validation/v1
kind: Validation
metadata:
  labels:
    foo: bar
```

which would fail on a disallowed apiVersion/kind, and a missing required "spec" field. is that correct?

Reply:
this line does validate the test object against the `foo` property's schema, but it only does [standard metadata validation](https://github.com/kubernetes/kubernetes/blob/c369cf187ea765c0a2387f2b39abe6ed18c8e6a8/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/validation.go#L30-L32), user-defined value validation is not honored

off-topic: `s` is unused in this function: https://github.com/kubernetes/kubernetes/blob/c369cf187ea765c0a2387f2b39abe6ed18c8e6a8/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/validation.go#L78